cmd/cache_serve.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/pelicanplatform/pelican/metrics"
+	nsregistry "github.com/pelicanplatform/pelican/namespace-registry"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/xrootd"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type directorResponse struct {
+	Error string `json:"error"`
+}
+
+func makeRequest(url string, method string, data map[string]interface{}, headers map[string]string) ([]byte, error) {
+	payload, _ := json.Marshal(data)
+	req, err := http.NewRequest(method, url, bytes.NewBuffer(payload))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	for key, val := range headers {
+		req.Header.Set(key, val)
+	}
+
+	tr := config.GetTransport()
+	client := &http.Client{Transport: tr}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Check HTTP response -- should be 200, else something went wrong
+	body, _ := io.ReadAll(resp.Body)
+	if resp.StatusCode != 200 {
+		return body, errors.Errorf("The URL %s replied with status code %d", url, resp.StatusCode)
+	}
+
+	return body, nil
+}
+
+func periodicAdvertiseCache(prefix string, nsAds []director.NamespaceAd) error {
+	ticker := time.NewTicker(1 * time.Minute)
+	go func() {
+		err := advertiseCache(prefix, nsAds)
+		if err != nil {
+			log.Warningln("Vsvhr advertise failed:", err)
+			if err = metrics.SetComponentHealthStatus("federation", "critical", "Error advertising cache to federation"); err != nil {
+				log.Warningln("Failed to update internal component health status:", err)
+			}
+		} else if err = metrics.SetComponentHealthStatus("federation", "ok", ""); err != nil {
+			log.Warningln("Failed to update internal component health status:", err)
+		}
+
+		for {
+			<-ticker.C
+			err := advertiseCache(prefix, nsAds)
+			if err != nil {
+				log.Warningln("Cache advertise failed:", err)
+				if err = metrics.SetComponentHealthStatus("federation", "critical", "Error advertising origin to federation"); err != nil {
+					log.Warningln("Failed to update internal component health status:", err)
+				}
+			} else if err = metrics.SetComponentHealthStatus("federation", "ok", ""); err != nil {
+				log.Warningln("Failed to update internal component health status:", err)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func advertiseCache(prefix string, nsAds []director.NamespaceAd) error {
+	name := param.Xrootd_Sitename.GetString()
+	if name == "" {
+		return errors.New("Cache name isn't set")
+	}
+
+	// TODO: waiting on a different branch to merge origin URL generation
+	// The checkdefaults func that runs before the origin is served checks for and
+	// parses the originUrl, so it should be safe to just grab it as a string here.
+	originUrl := param.Origin_Url.GetString()
+
+	// TODO: Need to figure out where to get some of these values
+	//               so that they aren't hardcoded...
+	cAd := director.NamespaceAd{
```

Comment:
The namespace ad here doesn't make much sense, does it?  Can we skip it?

Reply:
What do you mean skip it? CacheAds are expected to advertise to the director with a list of namespaces they have access to. At least that's how it was explained to me.